### DISCUSSION TITLE
How to define JSON custom scalar type in graphql which can be nullable?

### DISCUSSION THREAD

**Discussion by Yifeng-Sigma** [graphql/graphql-js#3981] [Q&A]
https://github.com/graphql/graphql-js/issues/3532#issuecomment-1125126566
I want to define a JSON custom scalar type which can be nullable. 

after graphql 16, `serialize()` can not return null, so I'm getting GRAPHQL error {"errors":[{"message":"Expected JsonType.serialize(\"null\") to return non-nullable value, returned: null" error where JsonType is my customized scalar.

Wondering what's the recommended way to do this?

**Comment by MukundaKatta** [comment]
The v16 change requires `serialize` to return a non-null value. For a nullable JSON scalar, just return a sentinel on null input — you want the caller to receive `null`, which happens when the *field* resolver returns `null`, not when the scalar serializes one. So inside `serialize`, if you receive `null` throw (it should never be called with null if the field is nullable); the GraphQL executor skips serialize for null field values entirely.

**Comment by yudin-s** [comment]
Nullability should be modeled on the field type, not inside the scalar.

For example:

```graphql
type Example {
  metadata: JSON
  requiredMetadata: JSON!
}
```

`metadata: JSON` means the field may be null. If the resolved field value is actually `null`, GraphQL should complete it as null before calling your scalar serialization logic.

So the scalar implementation should focus on valid JSON values when it is called:

```ts
const JSONScalar = new GraphQLScalarType({
  name: 'JSON',
  serialize(value) {
    if (value === null || value === undefined) {
      throw new TypeError('JSON scalar cannot serialize null directly')
    }
    return value
  },
  parseValue(value) {
    return value
  },
})
```

If `serialize(null)` is being called, I would check the resolver return path and the field type. Common causes are returning the string `"null"` from storage, wrapping the scalar in custom code before GraphQL completes null values, or marking the field as non-null somewhere in the schema.

In short: `JSON` can be nullable as a field, but the scalar itself does not need to “support null” as a JSON value.
